From 4bb7ceead98ff873b212b73af2d7dfb09053fd9d Mon Sep 17 00:00:00 2001
From: mintoyatsu <53593903+mintoyatsu@users.noreply.github.com>
Date: Tue, 10 Mar 2020 03:40:34 -0700
Subject: [PATCH] Server difficulty setting

Allows difficulty selection using server.properties. The "difficulty"
property has four values: 0 (Peaceful), 1 (Easy), 2 (Normal), 3 (Hard).
"spawn-monsters" can still be used to disable monster spawning
regardless of difficulty.

diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index a57fbf77..733c742f 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -175,8 +175,8 @@ public abstract class EntityHuman extends EntityLiving {
     }
 
     public void v() {
-        // CraftBukkit - spawnMonsters -> allowMonsters
-        if (!this.world.allowMonsters && this.health < 20 && this.ticksLived % 20 * 12 == 0) {
+        if (this.world.spawnMonsters == 0 && this.health < 20 && this.ticksLived % 20 * 12 == 0) {
+            // CraftBukkit - added regain reason of "REGEN" for filtering purposes.
             this.b(1, RegainReason.REGEN);
         }
 
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 645430ae..60780f56 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -315,7 +315,7 @@ public class MinecraftServer implements Runnable, ICommandListener {
 
             world.tracker = new EntityTracker(this, dimension);
             world.addIWorldAccess(new WorldManager(this, world));
-            world.spawnMonsters = this.propertyManager.getBoolean("spawn-monsters", true) ? 1 : 0;
+            world.spawnMonsters = this.propertyManager.getInt("difficulty", 1); // Canyon - difficulty switch (0=peaceful, 1=easy, 2=normal, 3=hard)
             world.setSpawnFlags(this.propertyManager.getBoolean("spawn-monsters", true), this.spawnAnimals);
             this.worlds.add(world);
             this.serverConfigurationManager.setPlayerFileData(this.worlds.toArray(new WorldServer[0]));
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 9b69e680..9646e0ad 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -365,6 +365,7 @@ public final class CraftServer implements Server {
 
         boolean animals = config.getBoolean("spawn-animals", console.spawnAnimals);
         boolean monsters = config.getBoolean("spawn-monsters", console.worlds.get(0).spawnMonsters > 0);
+        int difficulty = config.getInt("difficulty", console.worlds.get(0).spawnMonsters); // Canyon - difficulty switch
 
         //console.onlineMode = config.getBoolean("online-mode", console.onlineMode); // Canyon - online mode doesn't work.
         console.spawnAnimals = config.getBoolean("spawn-animals", console.spawnAnimals);
@@ -372,7 +373,7 @@ public final class CraftServer implements Server {
         console.allowFlight = config.getBoolean("allow-flight", console.allowFlight);
 
         for (WorldServer world : console.worlds) {
-            world.spawnMonsters = monsters ? 1 : 0;
+            world.spawnMonsters = difficulty;  // Canyon - difficulty switch
             world.setSpawnFlags(monsters, animals);
         }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index c51e06fb..b9e8290b 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -34,6 +34,10 @@ import org.bukkit.World;
 import org.bukkit.block.Biome;
 import org.bukkit.generator.BlockPopulator;
 
+// Canyon start
+import org.bukkit.Difficulty;
+// Canyon end
+
 public class CraftWorld implements World {
     private final WorldServer world;
     private Environment environment;
@@ -590,6 +594,16 @@ public class CraftWorld implements World {
     }
     // Canyon End
 
+    // Canyon start - difficulty API
+    public void setDifficulty(Difficulty difficulty) {
+        this.getHandle().spawnMonsters = difficulty.getValue();
+    }
+
+    public Difficulty getDifficulty() {
+        return Difficulty.getByValue(this.getHandle().spawnMonsters);
+    }
+    // Canyon end
+
     public boolean hasStorm() {
         return world.worldData.hasStorm();
     }
-- 
2.26.2

